controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -320,51 +323,66 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
-
-        CompletableFuture<Void> notify = segmentsFuture.thenComposeAsync(activeSegments ->
-                notifyTxnCreation(scope, stream, activeSegments, txnId), executor).whenComplete((v, e) ->
-                // Method notifyTxnCreation ensures that notification completes
-                // even in the presence of n/w or segment store failures.
-                log.debug("Txn={}, notified segments stores", txnId));
-
-        // Step 5. Start tracking txn in timeout service
-        return notify.whenCompleteAsync((result, ex) -> {
-            int version = 0;
-            long executionExpiryTime = System.currentTimeMillis() + maxExecutionPeriod;
-            if (!txnFuture.isCompletedExceptionally()) {
-                version = txnFuture.join().getVersion();
-                executionExpiryTime = txnFuture.join().getMaxExecutionExpiryTime();
-            }
-            timeoutService.addTxn(scope, stream, txnId, version, lease, executionExpiryTime, scaleGracePeriod);
-            log.debug("Txn={}, added to timeout service on host={}", txnId, hostId);
-        }, executor).thenApplyAsync(v -> new ImmutablePair<>(txnFuture.join(), segmentsFuture.join()), executor);
+                    // Step 4. Notify segment stores about new txn.
+                    CompletableFuture<List<Segment>> segmentsFuture = txnFuture.thenComposeAsync(txnData ->
+                            streamMetadataStore.getActiveSegments(scope, stream, txnData.getEpoch(), ctx, executor), executor);
+
+                    CompletableFuture<Void> notify = segmentsFuture.thenComposeAsync(activeSegments ->
+                            notifyTxnCreation(scope, stream, activeSegments, txnId), executor).whenComplete((v, e) ->
+                            // Method notifyTxnCreation ensures that notification completes
+                            // even in the presence of n/w or segment store failures.
+                            log.debug("Txn={}, notified segments stores", txnId));
+
+                    // Step 5. Start tracking txn in timeout service
+                    return notify.whenCompleteAsync((result, ex) -> {
+                        int version = 0;
+                        long executionExpiryTime = System.currentTimeMillis() + maxExecutionPeriod;
+                        if (!txnFuture.isCompletedExceptionally()) {
+                            version = txnFuture.join().getVersion();
+                            executionExpiryTime = txnFuture.join().getMaxExecutionExpiryTime();
+                        }
+                        timeoutService.addTxn(scope, stream, txnId, version, lease, executionExpiryTime, scaleGracePeriod);
+                        log.debug("Txn={}, added to timeout service on host={}", txnId, hostId);
+                    }, executor).thenApplyAsync(v -> new ImmutablePair<>(txnFuture.join(), segmentsFuture.join()), executor);
+                }), e -> {
```

Comment:
It would be really nice if you could reformat this block of code. It's really hard to read as callbacks are not indented as they should. Check out where line 383 starts, in relation to the lines above them.

Reply:
i agree. i have kept the original code and just wrapped it in a retry block as part of the changes here (also added one additional step of generating the txn id).